SPECK/src/main/java/SPEck.java

```diff
@@ -303,85 +383,104 @@ public static String[] parallelGenDataset(int i, int T, double theta, String str
      */
-    void execute(int P, int T, int parallelization, double theta, double fwer) throws IOException, ClassNotFoundException, InterruptedException {
-        String file = fileIn.split("\\.")[0];
+    void execute(int P,
+                 int T, 
+                 int numCores, 
+                 double theta, 
+                 double fwer
+    ) throws IOException,
+             ClassNotFoundException, 
+             InterruptedException 
+    {
+        String file = obsDatasetPath.split("\\.")[0];
         String fileRandom = file + "_random.txt";
         String fileMinned = file + "_mined.txt";
         // loads the input dataset in the data structures provided in input
-        loadDataset(fileIn,positionOrigin,datasetOrigin);
+        loadDataset(obsDatasetPath,positionOrigin,datasetOrigin);
         // fixes the seed for the first random generator
         Random r = new Random();
         System.err.println("Generating t datasets...");
-        generateTDatasets(T, theta, parallelization);
+        generateTDatasets(T, theta, numCores);
+
         double[] minPvalue = new double[P];
         // for all the P datasets of the WY method
         System.err.println("Calculating p values for corrected threshold in spark...");
         for(int j=0;j<P;j++) {
-            datasetRandom = Utils.generateDataset(datasetOrigin, positionOrigin, strategy, r);
+            Utils.Dataset datasetRandom = Utils.generateDataset(datasetOrigin, positionOrigin, strategy, r);
             // writes the actual random dataset generated
             writeDataset(fileRandom,datasetRandom);
             // mines the actual random dataset generated
             mining(fileRandom, fileMinned, theta);
             // computes in parallel the p-values of the actual random dataset
-            int[] pValueInt = computePValues(T,parallelization,fileMinned);
+            int[] pValueInt = computePValues(T,numCores,fileMinned);
             // finds the minimum of the p-values computed from the actual random dataset
             int min = Integer.MAX_VALUE;
             for (int i : pValueInt) {
                 if (min > i) min = i;
             }
             minPvalue[j] = (1 + min) / (T * 1. + 1);
         }
+
         System.err.println("Combining spark results...");
         // sorts the P minimum p-values
         Arrays.sort(minPvalue);
         // computes the corrected threshold
         correctedThreshold = minPvalue[(int)(P*fwer)];
         // if the corrected threshold is greater than the minimum possible value
+
         if(correctedThreshold!=1/(T * 1. + 1)){
             System.err.println("Calculating final p-values...");
             // mines the starting dataset
-            mining(fileIn, fileMinned, theta);
-            file = fileIn;
+            mining(obsDatasetPath, fileMinned, theta);
+            file = obsDatasetPath;
+            
             // computes the p-values of the starting dataset in parallel
```

Comment:
Update to "computes the number of the T datasets with results more or as extreme than in the observed one"